crates/hir_ty/src/tests/method_resolution.rs

```diff
@@ -1314,3 +1314,38 @@ fn f() {
     "#,
     );
 }
+
+#[test]
+fn inherent_impl_in_unnamed_const() {
+    check_types(
+        r#"
+struct S;
+
+const _: () = {
+    impl S {
+        fn method(&self) -> u16 { 0 }
+
+        pub(super) fn super_method(&self) -> u16 { 0 }
+
+        pub(crate) fn crate_method(&self) -> u16 { 0 }
+
+        pub fn pub_method(&self) -> u16 { 0 }
+    }
+};
+
+fn f() {
+    S.method();
+  //^^^^^^^^^^ u16
```

Comment:
Showerthought: once we start producing diagnostics for type mismatches, we should be able to write this test as

```rust
let _: u16 = S.method();
```

Reply:
we already record mismatches, so we could have a `check_no_mismatches` right now

Reply:
I wonder if it makes to migrate our test suite to that style? The specific benefit is that we'd also test for unrelated false positives. It also feels neater to not need the `//^^` annotations. 

On the other hand, annotations are helpful for seeing what exactly is being tested. And I think we can add "no diagnostics" condition to `//^^` tests just as well? 

Yet another considiration is that "no diagnostics" is not really robustly possible without having all of the lang items in, as some syntaxes depend on the lang items. 

Yeah, I am conflicted here. It feels relatively important to use the right approach for testing, but I am not sure which one is better....

Reply:
IMO `let _: u32 = ...` is pretty clear as to what is being tested, and having both would be somewhat redundant.

We don't necessarily need to check "no diagnostics", just "no type mismatches" would be enough, though probably checking that there are no diagnostics produced by inference (when it does so) would also make sense. I would expect they can be avoided without problems.

Reply:
One point against `let _: u32 = ...`, though, is that it influences inference. You can't test type variable fallback that way...

Reply:
Plus when we introduce inference diagnostics, it might be nice to test _them_ with the `//^^` syntax...